### PR TITLE
[19.09] Fix detached parent instance error in 8769

### PR DESCRIPTION
**Pull request by mvdbeek**
It's sort of expected that using the `ToolShedRepositoryCache` we can run into the situation where the original session from which the `ToolShedRepository` objects are loaded can be closed.

We normally work around that case in https://github.com/galaxyproject/galaxy/blob/7c5e72a7443d85ef50990f44fec2e173433cbb4f/lib/galaxy/tools/cache.py#L185

https://github.com/galaxyproject/galaxy/issues/8769 occurs when accessing attributes of `ToolShedRepository` objects, so I assume that's why the `except DetachedInstanceError` try/except logic doesn't work.

The workaround is to eagerload all `ToolShedRepository` attributes up front. I think medium term I'd like to redesign the whole install db system.

### CODE REVIEW

**Comment by nsoranzo** [comment]
Can we get rid of the `rebuild()` call then?

**Comment by mvdbeek** [comment]
Yeah, we can probably get rid of it. We still need to rebuild the cache occasionally, but I'd assume we can't run into the exception anymore.

File: lib/galaxy/tools/cache.py [hunk 1]
```diff
@@ -173,24 +166,7 @@ def rebuild(self):
...
-            self.rebuild()
-            return self._get_installed_repository(tool_shed=tool_shed,
-                                                  name=name,
-                                                  owner=owner,
-                                                  installed_changeset_revision=installed_changeset_revision,
-                                                  changeset_revision=changeset_revision,
-                                                  repository_id=repository_id)
-
-    def _get_installed_repository(self, tool_shed=None, name=None, owner=None, installed_changeset_revision=None, changeset_revision=None, repository_id=None):
+     def get_installed_repository(self, tool_shed=None, name=None, owner=None, installed_changeset_revision=None, changeset_revision=None, repository_id=None):
```
Comment (nsoranzo) [inline lib/galaxy/tools/cache.py]: Indentation is 5 spaces here.

Comment (mvdbeek) [reply lib/galaxy/tools/cache.py]: Oops, thanks!

**Review by nsoranzo — APPROVED** [review]

**Comment by martenson** [comment]
I am not a fan of the second commit going to release. Could we move that one to dev please?

**Comment by mvdbeek** [comment]
sure, can you cherry-pick and rebase this? I think eu needs this asap and I'm about to change trains.

**Comment by martenson** [comment]
sure thing

**Comment by bgruening** [comment]
I think we need to revert this one. It increases the memory usage of handler 4 fold.

grafik

handler3 is running the reverted Galaxy codebase. Handler 0,1,2 are running this patch. As you can see the memory consumption is 4 times higher :(

Sorry!

**Pull request merged**